File: java/tutorial/tut5-shoot_kill_ai/src/infinity/GameObject.java
package infinity;

import infinity.gameengine.Vector2;

import java.awt.Graphics2D;


public abstract class GameObject
    {
        private static int gameObjectIdCount = 0;

                private  int gameObjectId;

        public int getGameObjectId()
        {
             return gameObjectId;
        }

        protected Vector2 CameraPosition()
        {
            return gameWorld.getCameraPosition();
        }

        protected Vector2 ScreenCenter()
        {
             return gameWorld.getScreenCenter();
        }

        protected  GameWorld gameWorld;
        protected Vector2 gamePosition;

        protected GameObject(GameWorld gameWorld, Vector2 gamePosition)
        {
            this.gameWorld = gameWorld;
            this.gamePosition = gamePosition;
            gameObjectId = gameObjectIdCount++;
        }

        public abstract void Update();

        public abstract void Draw(Graphics2D g);//SpriteBatch spriteBatch);

        protected Vector2 PositionOnScreen()
        {
            return new Vector2().add(ScreenCenter()).subtract(CameraPosition()).add(gamePosition);
        }

        public boolean IsAlive()
        {
             return true;
        }
        
    }

File: src/Infinity/Infinity/GameObject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.Graphics;

namespace Infinity
{
    public abstract class GameObject
    {
        private static int gameObjectIdCount = 0;

        public int GameObjectId
        {
            get; private set;
        }

        protected Vector2 CameraPosition
        {
            get { return gameWorld.CameraPosition; }
        }

        protected Vector2 ScreenCenter
        {
            get { return gameWorld.ScreenCenter; }
        }

        protected readonly GameWorld gameWorld;
        protected Vector2 gamePosition;

        protected GameObject(GameWorld gameWorld, Vector2 gamePosition)
        {
            this.gameWorld = gameWorld;
            this.gamePosition = gamePosition;
            GameObjectId = gameObjectIdCount++;
        }

        public virtual void Update()
        {
            
        }

        public abstract void Draw(SpriteBatch spriteBatch);

        protected Vector2 PositionOnScreen
        {
            get { return ScreenCenter - CameraPosition + gamePosition; ; }
        }

        public virtual bool IsAlive
        {
            get { return true; }
        }
    }
}